revised - type specific handling of channels

J0EK3R opened the pull request:
Please have a look...

Now different channel types are possible.
I.e. the "shot canon" function of the tank on MK5.Channel3 is handled diffently then the other channels.
The virtual MK5.Channel5 handles the "turret lock option".

```
protected override (int, Func<float, (byte, bool)>) CreateChannelHandler(int channelNo)
    {
        return channelNo switch
        {
            0 => (0, SetOutput_AnalogChannel),                    // Tracks A + B forward, backward
            1 => (1, SetOutput_AnalogChannel_Turret),             // Turret rotation
            2 => (2, SetOutput_Shot),                             // shot canon
            3 => (3, SetOutput_AnalogChannel),                    // turn on spot
            4 => (VIRTUALCHANNEL, SetOutput_Option_Turret_Lock),  // virtual channel: lock turret
            _ => throw new ArgumentException("Illegal Argument", nameof(channelNo))
        };
    }
```

Commit:
JK: revised - channel types possible

Commit:
JK: added virtual channel handling

Commit:
JK: removed unnecessary call

Commit:
JK: fixed bad typo :|

Commit:
JK: revised SetOutput

Commit:
JK: fixed if condition

J0EK3R commented:
Oh, rebasing is needed - I'll do that tomorrow

Commit:
JK: clean up

vicocz commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MK4.cs:
```diff
@@ -2,136 +2,166 @@
+            Device1 => 0,
+            Device2 => 1,
+            Device3 => 2,
+            _ => throw new ArgumentException("Illegal Argument", nameof(address))
```
Nit pick: message should contain the value.

J0EK3R replied:
OK

https://github.com/J0EK3R/brickcontroller2/commit/2210a51d59bc7ad6573e2c5786bc223a2c7e6916

vicocz commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MKBaseNibble.cs:
```diff
@@ -41,24 +42,32 @@ protected enum ChannelType
+    /// containing the baseTelegramChannelByteOffset, the channel specific setvalue and a boolean to indicate a zero value.
     /// </summary>
-    protected readonly int _channelStartOffset;
+    protected readonly Func<float, bool>[] _setChannel;
```
I dislike this way to store func here. Is there a way to call virtual method instead on the affected places?

J0EK3R replied:
OK?
But it's just an array...
Simple, fast...
Any better suggestion?

vicocz replied:
I had to clone your fork and check the branch. Will let you know.

vicocz replied:
In base class add method which is refactoring of your's:
```
    protected bool SetChannelOutput(int channelNo, float value)
    {
        //return _setChannel[channelNo](value);
        if (IsVirtualChannel(channelNo))
        {
            // virtual channel
            (byte _, bool isModified) = ProccessChannelValue(channelNo, value);
            return isModified;
        }
        // real channel
        bool isOdd = (channelNo & 0x01) == 0x01;
        int byteOffset = GetByteOffset(channelNo);
        int specificChannelNo = GetSpecificChannelNumber(channelNo);

        (byte setValue_nibble, bool zeroSet) = ProccessChannelValue(channelNo, value);

        _bluetoothAdvertisingDeviceHandler.SetChannelState(specificChannelNo, zeroSet); // set global channel state
        return SetChannelValue(byteOffset, isOdd, setValue_nibble);
    }
```

requires some abstraction:
```
    protected virtual bool IsVirtualChannel(int channelNo) => false;

    protected abstract (byte value, bool flag) ProccessChannelValue(int channelNo, float value);
```

E.g. in MK5:
```
    protected override bool IsVirtualChannel(int channelNo) => channelNo == 4;
```

```
    protected override (byte value, bool flag) ProccessChannelValue(int channelNo, float value) => channelNo switch
    {
        0 => SetOutput_AnalogChannel(value),                    // Tracks A + B forward, backward
        1 => SetOutput_AnalogChannel_Turret(value),             // Turret rotation
        2 => SetOutput_Shot(value),                             // shot canon
        3 => SetOutput_AnalogChannel(value),                    // turn on spot
        4 => SetOutput_Option_Turret_Lock(value),               // virtual channel: lock turret
        _ => throw new ArgumentException("Illegal Argument", nameof(channelNo))
    };
```

vicocz replied:
Something like that, did not test it properly, just to smoke it works, I guess you need to change yet:
```
        valueChanged |= SetChannelOutput(1, _storedValues[1]); // The turret lock is a virtual channel, so call handler for real turret channel

```

J0EK3R replied:
double posted https://github.com/vicocz/brickcontroller2/pull/149#discussion_r2169600897

J0EK3R replied:
OK, it's in

https://github.com/J0EK3R/brickcontroller2/commit/f97a098c8051d18759eb0e6a34e2ad70180f908d

vicocz commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MK5.cs:
```diff
@@ -39,49 +44,209 @@ public MK5(string name, string address, byte[] deviceData, IDeviceRepository dev
+    /// <item><description>Channel 5: virtual channel for locking the turret</description></item>
+    /// </list></remarks>
+    /// </summary>
+    public override int NumberOfChannels => 5;
```
I'm OK with having virtual channel, just need to look how it behaves and review that.

J0EK3R replied:
Yes, please.
Btw: I like it! ;)

vicocz commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDeviceHandler.cs:
```diff
@@ -109,30 +109,63 @@ public BluetoothAdvertisingDeviceHandler(IBluetoothLEService bleService, ushort
+                }
+                else
+                {
+                    return false; // no change in state
```
Nit pick: I prefer to have single `return false` to simplify if/else branching.

J0EK3R replied:
OK, I've changed the code

https://github.com/J0EK3R/brickcontroller2/commit/f2a8058f8d786dc4226ef1bae2d15f3748cb196f

J0EK3R commented:
Btw.: did you recognize the target branch of this PR - it's your "local/mk5".

Commit:
JK: message of ArgumentException contains illegal value

Commit:
JK: copilot's suggestion

[nitpick] Variable 'realChannelNo' shadows the tuple element from setChannelHandler; consider deconstructing the tuple directly (e.g., 'var (realChannelNo, handler) = CreateChannelHandler(channelNo)') to avoid confusion.

J0EK3R commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MKBaseNibble.cs:
```diff
@@ -67,193 +76,141 @@ protected MKBaseNibble(string name, string address, byte[] deviceData, IDeviceRe
-                break;
-            case ChannelType.Left_Off_Right:
+            int realChannelNo = setChannelHandler.Item1;
 
```
OK, I've done this

https://github.com/J0EK3R/brickcontroller2/commit/93533c1ae1182af021f7fdd50d933d2fb34c3c55

Commit:
JK: lock to ensure ensure thread safety when updating the bitfield

J0EK3R commented on BrickController2/BrickController2/DeviceManagement/BluetoothAdvertisingDeviceHandler.cs:
```diff
@@ -109,30 +109,63 @@ public BluetoothAdvertisingDeviceHandler(IBluetoothLEService bleService, ushort
+        /// the specified channel.</param>
+        /// <returns><see langword="true"/> if the state of all channels is zero and the state of the specified channel was
+        /// successfully changed; otherwise, <see langword="false"/>.</returns>
+        public bool SetChannelState(int specificChannelNo, bool zeroSet)
```
OK, done

https://github.com/J0EK3R/brickcontroller2/commit/f2a8058f8d786dc4226ef1bae2d15f3748cb196f

vicocz commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MK5.cs:
```diff
@@ -28,8 +26,15 @@ internal class MK5 : MKBaseNibble, IDeviceType<MK5>
+    /// * 0x00: turret is unlocked
+    /// * 0x02: turret is locked
+    /// </summary>
+    private byte _turret_lock_Nibble = 0x00;
```
This variable should be reset after reconnection.

J0EK3R replied:
OK, I'll reset it on connect

https://github.com/J0EK3R/brickcontroller2/commit/5e5fef77829a64d5a042dcd6e85cdbaba911dbc7

vicocz commented:
> Btw.: did you recognize the target branch of this PR - it's your "local/mk5".

Yes I do, but unless I connect your for as remote origin, I need to clone your fork to see your branch in VS.

vicocz commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MKBaseNibble.cs:
```diff
@@ -41,219 +42,173 @@ protected enum ChannelType
+    /// array to hold the incoming output values for all channels.
     /// </summary>
-    protected readonly int _channelStartOffset;
+    protected readonly float[] _storedValues;
```
Is it reset on reconnect?

J0EK3R replied:
Now, it's done.

https://github.com/J0EK3R/brickcontroller2/commit/bc57da4d163a6270add7357947b5c17ffe56cd6e

J0EK3R commented:
OK, I've taken a look at your suggestion: it's essentially the same code; the main difference is that channel handling is decided at runtime each time the "SetOutput" method is called, rather than once in the constructor (-> "CreateSetChannelList") – the old game of runtime versus memory usage.

The question is, which is better? Do you have any technical arguments for your suggestion?

~~...I'll try to switch to your suggestion.~~

I've switched to your suggestion

https://github.com/J0EK3R/brickcontroller2/commit/f97a098c8051d18759eb0e6a34e2ad70180f908d

Commit:
JK: reset option _turret_lock_Nibble on connect

Commit:
JK: reset _stored Values on InitDevice

Commit:
JK: switched to Vit's suggestion

Commit:
JK: implemented DisconnectDevice to enshure valid channel states in _bluetoothAdvertisingDeviceHandler
